[Design] Introduce GCP Secrets for Gen AI services (#1697)
theopenconversationkit/tock: docs/_en/dev/feature-technical-designs/1696-gcp-secrets.md

---
title: GCP - Overall secret management design
---

# GCP - Overall secret management design #1696

- Proposal PR: [https://github.com/theopenconversationkit/tock/pull/1697](https://github.com/theopenconversationkit/tock/pull/1697)
- Github Issue for this feature: https://github.com/theopenconversationkit/tock/issues/1696

API version used : v1 of Google Secret Manager, reference version here.

## Introduction

We have already an integration of AWS Secret manager essentially used for :

> * **Generative AI services** dynamic secrets management, secrets vault are create dynamically for generative AI services (LLM, Embedding, Observability, VectorDB) 

Reviewer: ```suggestion
* **Generative AI services** dynamic secrets management, secrets vault are created dynamically for generative AI services (LLM, Embedding, Observability, VectorDB) 
```

Author: Added here  b2d463e3b

* **MongoDB credentials** : Add an implementation of `ai.tock.shared.security.mongo.MongoCredentialsProvider` ([MongoCredentialsProvider](https://github.com/theopenconversationkit/tock/blob/tock-24.3.4/shared/src/main/kotlin/security/mongo/MongoCredentialsProvider.kt#L24)) for GCP. Currently we only have the connexion string (thought environment variables), but credentials can be provided using ([MongoCredentialsProvider](https://github.com/theopenconversationkit/tock/blob/tock-24.3.4/shared/src/main/kotlin/security/mongo/MongoCredentialsProvider.kt#L24)) (see [MongoClient creation here](https://github.com/theopenconversationkit/tock/blob/tock-24.3.4/shared/src/main/kotlin/Mongos.kt#L178), [asyncMongoClient here](https://github.com/theopenconversationkit/tock/blob/tock-24.3.4/shared/src/main/kotlin/Mongos.kt#L197))
* **iAdvize Connector GraphQL credentials** : currently we have an AWS Secret Manager implementation of [`ai.tock.shared.security.credentials.CredentialsProvider`](https://github.com/theopenconversationkit/tock/blob/tock-24.3.4/shared/src/main/kotlin/security/credentials/CredentialsProvider.kt#L19) this will need to be implemented also for GCP Secret Manager. 

> Each of these 3 differents secret usages can use different providers idependently, for instance you could provide your mongodb credential using environment variables and handle the generate ai related secrets using GCP Secret Manager.

Author: @AurelienSylvan we have decided with @assouktim that secret providers should be independant for each scopes : gen-ai secrets, database (mongodb) secrets, iadvize secrets

In our case we will have : 
- gen ai secrets using GCP Secret Manager
- mongodb secrets / credentials using environment variable (kube secrets behind it)
- iAdvize secret using either environment variable (kube secret, we need to create it) or GCP secret ? ==> We need to decide for that point this will have role attribution impact if we use GCP Secret (see section [Role `tock.iadvize-secret.shareholder`](https://github.com/theopenconversationkit/tock/pull/1697/files#diff-846320cbce302e5bc02e2637c1343860db322ee9127551bd4ae715c10cd0d9ecR126) )

Author: We will use Kubescret (ENV) for iAdvize also.

> In this design we will as much as possible **refer to GCP Secrets using their secret name et not that full ressource name** that include the project-number (using `SECRET-NAME` instead of `projects/478160847739/secrets/SECRET-NAME`), we want to do this so that we can move ressources from a project to an other without having to do any kind of database data migration (changing project-number in some stored configuration will be difficult). Nevertheless applications needs to be aware of the GCP project number where the secrets are located, they will all rely on `tock_gcp_secret_project_number` environment variable to do so (see last section about environment variables for more details).

Reviewer: Project number or project id ? I thought we agreed on using the latter

  * trim / cut at 255 characters

### Secret `FUNCTION` part

| Description | AWS Secret name | GCP Secret name |
| --- | --- | --- |
| Observability tool secret. <br><br> *Eg. Langfuse secret key.* | `OBSERVABILITY/observabilitySetting` | `OBSERVABILITY-observabilitySetting` |
| RAG Embedding APIs secret. <br><br> *Eg. OpenAI Ada secret key.*  | `RAG/embeddingQuestion` | `RAG-embeddingQuestion` |
| RAG LLM Answer generation APIs secret <br><br> *Eg. Open AI LLM key.*  | `RAG/questionAnswering` | `RAG-questionAnswering` |
| Sentence Generation LLM API secret <br><br> *Eg. Open AI LLM key.*| `COMPLETION/sentenceGeneration` | `COMPLETION-sentenceGeneration` |

Exemple for bot named `my_bot` with namespace `my_namespace` under the GCP with number `478160847739` :
 * Rag Setting LLM secret key will be stored on GCP under the following :
    * Secret Name : PROD-TOCK-**my_namespace**-**my_bot**-GenAI-RAG-questionAnswering
    * Secret ressource (full) name : projects/**478160847739**/secrets/PROD-TOCK-**my_namespace**-**my_bot**-GenAI-RAG-questionAnswering
 * Generate sentence LLM secret key will be stored on GCP under the following :

>     * Secret Name : PROD-TOCK-**my_namespace**-**my_bot**-GenAI-RAG-questionAnswering

Reviewer: ```suggestion
    * Secret Name : PROD-TOCK-**my_namespace**-**my_bot**-GenAI-COMPLETION-sentenceGeneration
```

Author: Fixed here : 128b35125

* Bot-Admin / studio
* Bot API
* NLP API

*This secret isn't set though TOCK, should be set using the GCP console or infrastructure scripting.*

### For iAdvize credentials secrets

Role `tock.iadvize-secret.shareholder` (reader) :
* secretmanager.versions.access
* secretmanager.versions.get

This role `tock.iadvize-secret.shareholder`, should be applied to the following stack components :
* Bot API

*This secret isn't set though TOCK, should be set using the GCP console or infrastructure scripting.*

## Secret formats

### Gen AI Orchestrator secret payload format

Secret payload data for all generative AI related secret have the following base format :

```json
{
    "secret":"xxxxxxxxxxxxxxxxxxxxxxxx"
}
```

### MongoDB Secret format

The mongo database secret format is the following :
```json
{
  "password": "********",
  "username": "myUser"
}
```

### iAdvize Secret payload format

The iAdvize secret payload data is the following one :
```json
{
  "password":"****************",
  "username":"myiAdvizeTechnicalMail@exemple.com"
}
```

This is the iAdvize technical account you use for GraphQL calls.

## Generative AI - database storage format

> Here is a exemple of how it should be stored in the database :

Reviewer: ```suggestion
Here is an exemple of how it should be stored in the database :
```

        "apiVersion": "2024-03-01-preview",
        "provider": "AzureOpenAIService"
    }
}
```

## Gen AI Orchestrator secret format in API calls

POST /llm-providers/OpenAI/setting/status
```json
{
  "setting": {
    "provider": "OpenAI",
    "api_key": {
      "type": "GcpSecretManager",
      "secretName": "PROD-TOCK-my_namespace-my_bot-GenAI-RAG-questionAnswering"
    },
    "temperature": 1.2,
    "prompt": "How to learn to ride a bike without wheels!",
    "model": "gpt-3.5-turbo"
  },
  "observability_setting": {}
}
```

## Environnement variable settings for secrets

This design introduce the new provider type `GcpSecretManager`.

### Bot Admin / Studio

|Environment variable name | Default | Allowed values | Description |
|--- |--- |--- |--- |
| `tock_database_credentials_provider`| `Raw` | `Raw`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store database (mongodb) access secrets : <br> <ul> <li>Raw: Store secret directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |

> | `tock_iadvize_credentials_provider`| `Raw` | `Raw`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store iAdvize secrets : <br> <ul> <li>Raw: Store secret directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |

Author: Why is the Bot Admin calling iAdvize ? to check the distribution rule ? We also need to update the architecture (networking diagram) for it.

> | `tock_iadvize_credentials_provider`| `Raw` | `Raw`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store iAdvize secrets : <br> <ul> <li>Raw: Store secret directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |

Reviewer 2: Oh no, you're right, we don't need it here. It's just used on the Api Bot.

| `tock_gen-ai_credentials_provider`| `Raw` | `Raw`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store generative AI secrets : <br> <ul> <li>Raw: Store secret directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |
| `tock_gen_ai_orchestrator_secret_storage_prefix_name`| `DEV` | any string ? | See section "Environment isolation `SECRET_STORAGE_PREFIX` and feature based environements" of this document.  <br><br> ⚠️ Current default value need to be changed. |
| `tock_mongodb_credentials_secret_name`| `PROD-TOCK-MONGODB` | GCP Secret name Path | Secret ID use for Mongo DB secret. Only if not passed in mongo URI using the `tock_mongo_url`. You should also include the GCP mongodb secret module to use it. |

> | `tock_gcp_secret_project_number`| `478160847739` | GCP project number | The GCP project number where secrets are stored. |

Author: Fixed here : 498d296ec

| `tock_iadvize_credentials_provider`| `Raw` | `Raw`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store iAdvize secrets : <br> <ul> <li>Raw: Store secret directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |
| `tock_gen-ai_credentials_provider`| `Raw` | `Raw`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store generative AI secrets : <br> <ul> <li>Raw: Store secret directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |
| `tock_mongodb_credentials_secret_name`| `PROD-TOCK-MONGODB` | GCP Secret name Path | Secret ID use for Mongo DB secret. Only if not passed in mongo URI using the `tock_mongo_url`. You should also include the GCP mongodb secret module to use it. |

> | `tock_gcp_secret_project_number`| `478160847739` | GCP project number | The GCP project number where secrets are stored. |

Reviewer: Same remark, shouldn't we use project id ?

Author: Removed from NLP-api, copy/paste mistake.
Environment variable renamed also to `tock_gcp_secret_project_id`

### Bot API

> |Environment variable name | Default | Allowed values | Description |

Author: tock_database_credentials_provider is ok, other env variables seems also ok, double check if you want

| `tock_mongodb_credentials_secret_name`| `PROD-TOCK-MONGODB` | GCP Secret name Path | Secret name use for Mongo DB secret. Only if not passed in mongo URI using the `tock_mongo_url`. You should also include the GCP mongodb secret module to use it. |
| `tock_gcp_secret_project_number`| `478160847739` | GCP project number | The GCP project number where secrets are stored. |
| `tock_gcp_iadvize_credentials_secret_name`| `PROD-TOCK-IADVIZE` | GCP Secret ID Path | Secret name use for Mongo DB secret iAdvize GraphQL calls. Only needed if you use iAdvize, it can also be passed though environment variables `iadvize_username_authentication` and `iadvize_password_authentication`. | 

### Gen AI Orchestrator

|Environment variable name | Default | Allowed values | Description |
|--- |--- |--- |--- |
| `tock_gcp_secret_project_number`| `478160847739` | GCP project number | The GCP project number where secrets are stored. |

## Technical change that should be made

### Breaking changes

> * Default value of `tock_gen_ai_orchestrator_secret_storage_prefix_name` currently `/dev` shouldn't use slashes as it's not allow according to GCP Secret Names constraints. It will be changed to `dev` but it might break all running project that doesn't defined it.

Reviewer 2: The ‘break’ is not linked to the presence of the slash, as the prefix for GCP is also standardised. However, it is linked to the default value of the prefix, which has been changed to ‘LOCAL/TOCK’.

The following is a list of the changes made to the existing environment variables: 
'tock_database_credentials_provider' replaced by 'tock_database_mongodb_secret_manager_provider'
'tock_iadvize_credentials_provider' replaced by 'tock_iadvize_secret_manager_provider'
'aws_iadvize_credentials_secret_id' replaced by 'tock_iadvize_credentials_secret_name'
'tock_gen_ai_orchestrator_secret_storage_type' replaced by 'tock_gen_ai_secret_manager_provider'
'tock_gen_ai_orchestrator_secret_storage_prefix_name' replaced by 'tock_gen_ai_secret_prefix'

Author: Added at the end for @AurelienSylvan to be renamed / updated in helm charts also.
@sacquatella those environment variables will have impact on the Helm Charts they are all specified in this design document that will be available on the EN version of tock website, I'll create on issue on https://github.com/theopenconversationkit/tock-helm-chart so that we don't forget it.

Author: Issue created here : https://github.com/theopenconversationkit/tock-helm-chart/issues/2

| `tock_iadvize_secret_manager_provider`| `Env` | `Env`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store iAdvize secrets : <br> <ul> <li>Env: Store the secret directly in an environment variable.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |
| `tock_gen_ai_secret_manager_provider`| `Null` | `Env`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store generative AI secrets : <br> <ul> <li>Env: Store the secret directly in an environment variable</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> Default (Null): The secret will be stores directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.|
| `tock_gen_ai_secret_prefix`| `LOCAL/TOCK` | any string ? | See section "Environment isolation `SECRET_STORAGE_PREFIX` and feature based environements" of this document. |

> | `tock_database_mongodb_secret_manager_provider`| `PROD-TOCK-MONGODB` | GCP Secret name Path | Secret ID use for Mongo DB secret. Only if not passed in mongo URI using the `tock_mongo_url`. You should also include the GCP mongodb secret module to use it. |

Reviewer 2: ```suggestion
| `tock_database_mongodb_credentials_secret_name`| `PROD-TOCK-MONGODB` | GCP Secret name Path | Secret ID use for Mongo DB secret. Only if not passed in mongo URI using the `tock_mongo_url`. You should also include the GCP mongodb secret module to use it. |
```

| `tock_gcp_project_id`| `my-gcp-project-id` | GCP project ID | The GCP project ID where secrets are stored. |

### NLP API

|Environment variable name | Default | Allowed values | Description |
|--- |--- |--- |--- |
| `tock_database_mongodb_secret_manager_provider`| `Raw` | `Raw`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store database (mongodb) access secrets : <br> <ul> <li>Raw: Store secret directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |

> | `tock_mongodb_credentials_secret_name`| `PROD-TOCK-MONGODB` | GCP Secret name Path | Secret ID use for Mongo DB secret. Only if not passed in mongo URI using the `tock_mongo_url`. You should also include the GCP mongodb secret module to use it. |

Reviewer 2: ```suggestion
| `tock_database_mongodb_credentials_secret_name`| `PROD-TOCK-MONGODB` | GCP Secret name Path | Secret ID use for Mongo DB secret. Only if not passed in mongo URI using the `tock_mongo_url`. You should also include the GCP mongodb secret module to use it. |
```

| `tock_iadvize_secret_manager_provider`| `Raw` | `Raw`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store iAdvize secrets : <br> <ul> <li>Raw: Store secret directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |
| `tock_gen-ai_credentials_provider`| `Raw` | `Raw`, `AwsSecretsManager`, `GcpSecretManager` | Type of credential provider used to store generative AI secrets : <br> <ul> <li>Raw: Store secret directly into tock's mongodb in raw. Use it only for local dev purposed it's clearly unsafe.</li><li>AwsSecretsManager: rely on AWS Secret Manager.</li><li>GcpSecretManager: rely on GCP Secret Manager.</li></ul> |

> | `tock_mongodb_credentials_secret_name`| `PROD-TOCK-MONGODB` | GCP Secret name Path | Secret name use for Mongo DB secret. Only if not passed in mongo URI using the `tock_mongo_url`. You should also include the GCP mongodb secret module to use it. |

Reviewer 2: ```suggestion
| `tock_database_mongodb_credentials_secret_name`| `PROD-TOCK-MONGODB` | GCP Secret name Path | Secret name use for Mongo DB secret. Only if not passed in mongo URI using the `tock_mongo_url`. You should also include the GCP mongodb secret module to use it. |
```

* Secret name generation (normalization) should not be done at the DTO mapper level as it currently assume that we only use AWS Secret Manager :
  *  [BotRAGConfigurationDTO.kt#L60](https://github.com/theopenconversationkit/tock/blob/master/bot/admin/server/src/main/kotlin/model/BotRAGConfigurationDTO.kt#L60), [BotRAGConfigurationDTO.kt#L66](https://github.com/theopenconversationkit/tock/blob/master/bot/admin/server/src/main/kotlin/model/BotRAGConfigurationDTO.kt#L66)
  * BotObservabilityConfigurationDTO
  * BotSentenceGenerationDTO
  * .... vector DB settings ...

### List of existing environement variables that were renamed, may introduce breaking change if used

* `tock_iadvize_credentials_provider` replaced by `tock_iadvize_secret_manager_provider`
* `aws_iadvize_credentials_secret_id` replaced by `tock_iadvize_credentials_secret_name`
* `tock_gen_ai_orchestrator_secret_storage_type` replaced by `tock_gen_ai_secret_manager_provider`

> * `tock_gen_ai_orchestrator_secret_storage_prefix_name` replaced by `tock_gen_ai_secret_prefix`

Reviewer 2: Also replaced :
`tock_database_credentials_provider` replaced by `tock_database_mongodb_secret_manager_provider`
`tock_mongodb_credentials_secret_name` replaced by `tock_database_mongodb_credentials_secret_name`
